Tests/SessionTests.swift

```diff
@@ -1435,6 +1435,43 @@ class SessionTestCase: BaseTestCase {
         // Then
         assertErrorIsAFError(error) { XCTAssertTrue($0.isSessionDeinitializedError) }
     }
+
+    // MARK: Tests - Request Cancellation
+
+    func testThatSessionOnlyCallsResponseSerializerCompletionWhenCancellingInsideCompletion() {
+        // Given
+        let handler = RequestHandler()
+        let session = Session()
+
+        let expectation = self.expectation(description: "request should complete")
+        var response: DataResponse<Any>?
+        var completionCallCount = 0
+
+        // When
+        let request = session.request("https://httpbin.org/get", interceptor: handler)
+        request.validate()
+
+        request.responseJSON { resp in
+            request.cancel()
+
+            response = resp
+            completionCallCount += 1
+
+            DispatchQueue.main.after(0.01) { expectation.fulfill() }
```

Comment:
Why is this delay necessary? Is it an attempt to finish the internal race to set the cancellation state? Also, how sensitive is this delay? We've had issues in the past with static delays being too short to reliably pass in the test environment.

Reply:
Great question. The reason the delay is necessary is to allow the `responseJSON` completion closure to be called a second time before the expectation is fulfilled. Without the delay, we'd have a race condition where the assertion on `completionCallCount` would most likely pass the assertion test before the retry logic actually ran the `responseJSON` completion closure again. In all my testing locally, this test will fail if we do not ship the provided fix. So that is good at least.

I personally do not like these types of tests that require waits to see if bad things happen, but unfortunately, this is one of those types of cases. We need a slight delay before we move on with the assertions to make sure additional calls aren't occurring. Thankfully, this will never produce false-positives if CI runs this particular test much slower. It would perhaps miss a regression if the delay was too short, but it would not produce a false-positive failure.

Reply:
Thinking about it again, I think we can do without the delay as, if the completion did fire again, the fulfillment count of the expectation would go up and the test would fail. I think that would probably be a bit cleaner than the delay, if you think it would work.

Reply:
It's cleaner, but way more racey, which is why I added the delay. By adding a delay, we know for certain that the code would execute a second time on a local machine which would produce the failure. Without the delay, it generally won't fail, depends on how fast the test can get shut down. My goal here was to produce a test that reliably fails with the current design flaw. Then fix the code so the test reliably passes.